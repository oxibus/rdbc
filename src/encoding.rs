// https://github.com/hsivonen/recode_rs

use std::io::{Read, Write};
use std::str::from_utf8_mut;

use encoding_rs::*;

use crate::error::DbcError;

pub fn utf8_to_gbk(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "UTF-8", "GBK")
}

pub fn gbk_to_utf8(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "GBK", "UTF-8")
}

pub fn to_utf8(src_encoding_label: &str, src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, src_encoding_label, "UTF-8")
}

pub fn recode(
    src_data: &[u8],
    src_encoding_label: &str,
    dst_encoding_label: &str,
) -> Result<Vec<u8>, DbcError> {
    let src_encoding = get_encoding(Some(src_encoding_label.to_string()))?;
    let dst_encoding = get_encoding(Some(dst_encoding_label.to_string()))?;

    let mut decoder = src_encoding.new_decoder();
    let mut encoder = dst_encoding.new_encoder();

    let mut buf = std::io::Cursor::new(Vec::new());

    convert_via_utf8(
        &mut decoder,
        &mut encoder,
        &mut std::io::Cursor::new(src_data),
        &mut buf,
        true,
    )?;

    Ok(buf.into_inner())
}

pub fn get_encoding(opt: Option<String>) -> Result<&'static Encoding, DbcError> {
    match opt {
        None => Ok(UTF_8),
        Some(label) => match Encoding::for_label(label.as_bytes()) {
            None => Err(DbcError::InvalidEncodingLabel(label)),
            Some(encoding) => Ok(encoding),
        },
    }
}

pub fn convert_via_utf8(
    decoder: &mut Decoder,
    encoder: &mut Encoder,
    read: &mut dyn Read,
    write: &mut dyn Write,
    last: bool,
) -> Result<(), DbcError> {
    let mut input_buffer = [0u8; 2048];
    let mut intermediate_buffer_bytes = [0u8; 4096];
    let intermediate_buffer: &mut str = from_utf8_mut(&mut intermediate_buffer_bytes[..]).unwrap();
    let mut output_buffer = [0u8; 4096];
    let mut current_input_ended = false;
    while !current_input_ended {
        match read.read(&mut input_buffer) {
            Err(e) => {
                log::error!("Error reading input, error = {}", e);
                return Err(DbcError::EncodingReadInputError);
            }
            Ok(decoder_input_end) => {
                current_input_ended = decoder_input_end == 0;
                let input_ended = last && current_input_ended;
                let mut decoder_input_start = 0usize;
                loop {
                    let (decoder_result, decoder_read, decoder_written, _) = decoder.decode_to_str(
                        &input_buffer[decoder_input_start..decoder_input_end],
                        intermediate_buffer,
                        input_ended,
                    );
                    decoder_input_start += decoder_read;

                    let last_output = if input_ended {
                        match decoder_result {
                            CoderResult::InputEmpty => true,
                            CoderResult::OutputFull => false,
                        }
                    } else {
                        false
                    };

                    // Regardless of whether the intermediate buffer got full
                    // or the input buffer was exhausted, let's process what's
                    // in the intermediate buffer.

                    if encoder.encoding() == UTF_8 {
                        // If the target is UTF-8, optimize out the encoder.
                        if let Err(e) =
                            write.write_all(&intermediate_buffer.as_bytes()[..decoder_written])
                        {
                            log::error!("Error writing output, error = {}", e);
                            return Err(DbcError::EncodingWriteOutputError);
                        }
                    } else {
                        let mut encoder_input_start = 0usize;
                        loop {
                            let (encoder_result, encoder_read, encoder_written, _) = encoder
                                .encode_from_utf8(
                                    &intermediate_buffer[encoder_input_start..decoder_written],
                                    &mut output_buffer,
                                    last_output,
                                );
                            encoder_input_start += encoder_read;
                            if let Err(e) = write.write_all(&output_buffer[..encoder_written]) {
                                log::error!("Error writing output, error = {}", e);
                                return Err(DbcError::EncodingWriteOutputError);
                            }
                            match encoder_result {
                                CoderResult::InputEmpty => {
                                    break;
                                }
                                CoderResult::OutputFull => {
                                    continue;
                                }
                            }
                        }
                    }

                    // Now let's see if we should read again or process the
                    // rest of the current input buffer.
                    match decoder_result {
                        CoderResult::InputEmpty => {
                            break;
                        }
                        CoderResult::OutputFull => {
                            continue;
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
