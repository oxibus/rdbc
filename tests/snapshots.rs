#![cfg(feature = "serde")]

use std::fs;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot, with_settings};
use rrdbc::ast::network_ast::parse_dbc;
use rrdbc::encoding::decode_cp1252;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        ("dbc/dbc_from_cantools", "dbc-cantools"),
        ("dbc/mytest", "mytest"),
    ];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(&path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) {
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match parse_dbc(buffer.as_ref()) {
            Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
            Err(e) => {
                eprintln!("Failed to parse {file_name}.dbc: {e:?}");
                assert_debug_snapshot!(format!("!error___{file_name}"), e);
            }
        }
    } else {
        let error = format!("Failed to decode {file_name}.dbc as cp1252");
        eprintln!("{error}");
        assert_snapshot!(format!("!error___{file_name}"), error);
    }
}
