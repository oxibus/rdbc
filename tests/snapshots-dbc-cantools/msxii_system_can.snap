---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
bit_timing:
  value: ~
nodes:
  - BMS
value_tables: ~
messages:
  - header:
      id: 1025
      name: BATTERY_VT
      size: 6
      transmitter: BMS
    signals:
      - name: MODULE_TEMP_35
        multiplexer:
          multiplexer_signal: 35
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_34
        multiplexer:
          multiplexer_signal: 34
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_33
        multiplexer:
          multiplexer_signal: 33
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_32
        multiplexer:
          multiplexer_signal: 32
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_31
        multiplexer:
          multiplexer_signal: 31
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_30
        multiplexer:
          multiplexer_signal: 30
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_29
        multiplexer:
          multiplexer_signal: 29
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_28
        multiplexer:
          multiplexer_signal: 28
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_27
        multiplexer:
          multiplexer_signal: 27
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_26
        multiplexer:
          multiplexer_signal: 26
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_25
        multiplexer:
          multiplexer_signal: 25
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_24
        multiplexer:
          multiplexer_signal: 24
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_23
        multiplexer:
          multiplexer_signal: 23
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_22
        multiplexer:
          multiplexer_signal: 22
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_21
        multiplexer:
          multiplexer_signal: 21
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_20
        multiplexer:
          multiplexer_signal: 20
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_19
        multiplexer:
          multiplexer_signal: 19
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_18
        multiplexer:
          multiplexer_signal: 18
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_17
        multiplexer:
          multiplexer_signal: 17
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_16
        multiplexer:
          multiplexer_signal: 16
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_15
        multiplexer:
          multiplexer_signal: 15
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_14
        multiplexer:
          multiplexer_signal: 14
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_13
        multiplexer:
          multiplexer_signal: 13
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_12
        multiplexer:
          multiplexer_signal: 12
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_11
        multiplexer:
          multiplexer_signal: 11
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_10
        multiplexer:
          multiplexer_signal: 10
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_09
        multiplexer:
          multiplexer_signal: 9
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_08
        multiplexer:
          multiplexer_signal: 8
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_07
        multiplexer:
          multiplexer_signal: 7
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_06
        multiplexer:
          multiplexer_signal: 6
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_05
        multiplexer:
          multiplexer_signal: 5
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_04
        multiplexer:
          multiplexer_signal: 4
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_03
        multiplexer:
          multiplexer_signal: 3
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_02
        multiplexer:
          multiplexer_signal: 2
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_01
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_TEMP_00
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_35
        multiplexer:
          multiplexer_signal: 35
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_34
        multiplexer:
          multiplexer_signal: 34
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_33
        multiplexer:
          multiplexer_signal: 33
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_32
        multiplexer:
          multiplexer_signal: 32
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_31
        multiplexer:
          multiplexer_signal: 31
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_30
        multiplexer:
          multiplexer_signal: 30
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_29
        multiplexer:
          multiplexer_signal: 29
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_28
        multiplexer:
          multiplexer_signal: 28
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_27
        multiplexer:
          multiplexer_signal: 27
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_26
        multiplexer:
          multiplexer_signal: 26
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_25
        multiplexer:
          multiplexer_signal: 25
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_24
        multiplexer:
          multiplexer_signal: 24
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_23
        multiplexer:
          multiplexer_signal: 23
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_22
        multiplexer:
          multiplexer_signal: 22
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_21
        multiplexer:
          multiplexer_signal: 21
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_20
        multiplexer:
          multiplexer_signal: 20
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_19
        multiplexer:
          multiplexer_signal: 19
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_18
        multiplexer:
          multiplexer_signal: 18
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_17
        multiplexer:
          multiplexer_signal: 17
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_16
        multiplexer:
          multiplexer_signal: 16
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_15
        multiplexer:
          multiplexer_signal: 15
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_14
        multiplexer:
          multiplexer_signal: 14
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_13
        multiplexer:
          multiplexer_signal: 13
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_12
        multiplexer:
          multiplexer_signal: 12
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_11
        multiplexer:
          multiplexer_signal: 11
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_10
        multiplexer:
          multiplexer_signal: 10
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_09
        multiplexer:
          multiplexer_signal: 9
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_08
        multiplexer:
          multiplexer_signal: 8
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_07
        multiplexer:
          multiplexer_signal: 7
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_06
        multiplexer:
          multiplexer_signal: 6
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_05
        multiplexer:
          multiplexer_signal: 5
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_04
        multiplexer:
          multiplexer_signal: 4
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_03
        multiplexer:
          multiplexer_signal: 3
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_02
        multiplexer:
          multiplexer_signal: 2
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_01
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: MODULE_VOLTAGE_00
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: BATTERY_VT_INDEX
        multiplexer:
          multiplexer_signal: ~
          multiplexer_switch: ~
        start_bit: 0
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
env_vars: []
env_vars_data: []
comments: []
attribute_definitions: []
attribute_defaults: []
attribute_values: []
signal_value_descriptions: []
env_var_value_descriptions: []
